core/modules/replica/HttpIngestModule.cc

```diff
@@ -0,0 +1,1465 @@
+    // dependent metadata, such as replicas info
+    config->deleteDatabase(databaseInfo.name);
+
+    // This step is needed to get workers' Configuration in-sync with its
+    // persistent state.
+    unsigned int const workerResponseTimeoutSec = 60;
+    error = _reconfigureWorkers(databaseInfo, allWorkers, workerResponseTimeoutSec);
+    if (not error.empty()) {
+        sendError(resp, __func__, error);
+        return;
+    }
+
+    json result;
+    sendData(resp, result);
+}
+
+
+void HttpIngestModule::_addTable(qhttp::Request::Ptr const& req,
+                                 qhttp::Response::Ptr const& resp) {
+    debug(__func__);
+
+    auto const config = controller()->serviceProvider()->config();
+
+    HttpRequestBody body(req);
+    auto const database      = body.required<string>("database");
+    auto const table         = body.required<string>("table");
+    auto const isPartitioned = (bool)body.required<int>("is_partitioned");
+    auto const schema        = body.required<json>("schema");
+    auto const isDirector    = (bool)body.required<int>("is_director");
+    auto const directorKey   = body.optional<string>("director_key", "");
+    auto const chunkIdColName    = body.optional<string>("chunk_id_key", "");
+    auto const subChunkIdColName = body.optional<string>("sub_chunk_id_key", "");
+    auto const latitudeColName  = body.optional<string>("latitude_key",  "");
+    auto const longitudeColName = body.optional<string>("longitude_key", "");
+
+    debug(__func__, "database="      + database);
+    debug(__func__, "table="         + table);
+    debug(__func__, "isPartitioned=" + string(isPartitioned ? "1" : "0"));
+    debug(__func__, "schema="        + schema.dump());
+    debug(__func__, "isDirector="    + string(isDirector ? "1" : "0"));
+    debug(__func__, "directorKey="   + directorKey);
+    debug(__func__, "chunkIdColName="    + chunkIdColName);
+    debug(__func__, "subChunkIdColName=" + subChunkIdColName);
+    debug(__func__, "latitudeColName="  + latitudeColName);
+    debug(__func__, "longitudeColName=" + longitudeColName);
+
+    // Make sure the database is known and it's not PUBLISHED yet
+
+    auto databaseInfo = config->databaseInfo(database);
+    if (databaseInfo.isPublished) {
+       sendError(resp, __func__, "the database is already published");
+        return;
+    }
+
+    // Make sure the table doesn't exist in the Configuration
+
+    for (auto&& existingTable: databaseInfo.tables()) {
+        if (table == existingTable) {
+            sendError(resp, __func__, "table already exists");
+            return;
+        }
+    }
+
+    // Translate table schema
+
+    if (schema.is_null()) {
+        sendError(resp, __func__, "table schema is empty");
+        return;
+    }
+    if (not schema.is_array()) {
+        sendError(resp, __func__, "table schema is not defined as an array");
+        return;
+    }
+
+    list<pair<string,string>> columns;
```

Comment:
consider a struct or class, this is pretty deep nesting.

Reply:
From my point of view, this statement is BOTH incorrect and correct. It's correct because I should have used a struct (or a class) here. And it's incorrect in is part which is related to the motivation behind replacing `<pair<string,string>` with a specialized type. The "deep nesting" is not a problem here at all. A reason for the refactoring has a different nature. The expressive type power of C++11 (on which , I guess, we both agree) allows and encourages using abstractions for everything. This makes the code more robust and more readable (less documentation is needed). The ability of expressing "light-weight abstractions" in the code is one of the most fundamental design ideas behind C++. You may read on this directly from the source (the language author's Web site):
http://www.stroustrup.com/ETAPS-corrected-draft.pdf

Compare, for example two ways of expressing column definitions:
```
pair<string,string> coldef;
// I know the first is the name of a column
coldef.first = "Column A";
// And I know that the second is its type
coldef.second = "VARCHAR(255)";
```
A problem with that class `pair` is way to generic, so I would have to tell everyone what is `first` and what is `second` in all contexts I would be using this definitions.
However if I have a structure which would look (for example) like this:
```
struct ColDef {
    std::string name;
    std::string type;
};
```
Then it becomes perfectly clear "who is who" and "what is what".
Having said that, I agree to refactor the code.